### DISCUSSION TITLE
Storage for persistant containers?

### DISCUSSION THREAD

**Discussion by adamshand** [caprover/caprover#1913] [Q&A]
Hello,

I'm very happily running a single CapRover instance, but looking at setting up a couple more, so I can have multiple instances of images and failover in case a server dies.

For images that don't require persistent storage, this seems fairly straight forward.  But how are people doing this for images which require persistent storage?  I've been researching options and everything seems a bit messy unless you have an NFS server (or similar).  

Is there a best practice for this using some form of cheapish cloud storage?  If not, what's working well for people?

Thanks for any insight!

**Comment by githubsaturn** [comment]
Your understanding is 100% correct. Distributed filesystem is a extremely complex problem. As a matter of fact, the perfect solution is [impractical](https://en.wikipedia.org/wiki/CAP_theorem). However, there are a few solutions available for these use cases:
- NFS
- Gluster
- Ceph
- And probably more..

**Reply by adamshand** [reply]
Stink, I was hoping somebody would have already figured out something that makes sense at the homelab level.  I don't really want to mess with k8s, ceph etc.

I think (?!) I can make this work using the rclone plugin for Docker and have containers sync their volume when they start/stop (and periodically).  Not sure if I'll be able to integrate that with CapRover or how well it will work in practice .... but will have a play and report back.

**Reply by githubsaturn** [reply]
Yea keep in mind that you have to shutdown your container before starting the backup process or else you risk damaging your files.

**Reply by adamshand** [reply]
Yep, aware of that.  Thanks.

Also, in case anyone finds this thread while searching, I'd missed this page in the documentation.

https://caprover.com/docs/stateless-with-persistent-data.html
